### ISSUE TITLE
[Documentation] Breathe doxygenstruct and doxygenclass fail in some cases 

### ISSUE THREAD

**Issue by arnfinn** [PCMSolver/pcmsolver#100]
Can not find the classes "Input", "Sphere", "ChargeDistribution", "Molecule" and "Meddle".

## Current Behavior

The following warnings are produced during build, and what is shown in the final documentation page:
```
/home/docs/checkouts/readthedocs.org/user_builds/pcmsolver/checkouts/rtd_testing/doc/code-reference/helper-classes.rst:10: WARNING: doxygenclass: Cannot find class "Input" in doxygen xml output for project "PCMSolver" from directory: xml
/home/docs/checkouts/readthedocs.org/user_builds/pcmsolver/checkouts/rtd_testing/doc/code-reference/helper-classes.rst:18: WARNING: doxygenstruct: Cannot find class "Sphere" in doxygen xml output for project "PCMSolver" from directory: xml
/home/docs/checkouts/readthedocs.org/user_builds/pcmsolver/checkouts/rtd_testing/doc/code-reference/helper-classes.rst:34: WARNING: doxygenstruct: Cannot find class "ChargeDistribution" in doxygen xml output for project "PCMSolver" from directory: xml
/home/docs/checkouts/readthedocs.org/user_builds/pcmsolver/checkouts/rtd_testing/doc/code-reference/helper-classes.rst:42: WARNING: doxygenclass: Cannot find class "Molecule" in doxygen xml output for project "PCMSolver" from directory: xml
/home/docs/checkouts/readthedocs.org/user_builds/pcmsolver/checkouts/rtd_testing/doc/users/interfacing.rst:60: WARNING: doxygenclass: Cannot find class "Meddle" in doxygen xml output for project "PCMSolver" from directory: xml
```
This is as shown [here](https://github.com/michaeljones/breathe/blob/master/documentation/source/class.rst#failing-example) (Breathe documentation, example of intentionally fails)

## Steps to Reproduce (for bugs)

1. `./setup.py`
2. `cd build`
3. `make doc`

**Comment by robertodr** [comment]
Can you grep in the Doxygen XML output to see if these classes are there? Might be that they are excluded from parsing in `Doxyfile.in` (found under `doc`) for some weird reason :/

**Comment by arnfinn** [comment]
OK, it seems like there is two ways to Rome. Either
- replace `.. doxygenclass:: Molecule` etc. with `.. doxygenclass:: pcm::Molecule`in the rst-files, or
- move the Doxygen section out of the `namespace pcm {` and to the head of the file.
```
/*! \file Molecule.hpp
 *  \class Molecule
...
*/
```

@robertodr: what is the proper way?

**Comment by robertodr** [comment]
Ugh! I didn't know that... So it might very well be that I broke docs when I introduced namespaces all over in the code... Anyhow, I think the best strategy is to modify all sources to have Doxygen groups for namespaces and then class documentation as usual. As an example:
```
/*!
 *  \addtogroup pcm
 *  @{
 */

//! Top-level PCMSolver namespace
namespace pcm {
using utils::Atom;
using utils::Sphere;

enum rotorType { rtAsymmetric, rtSymmetric, rtSpherical, rtLinear, rtAtom };
const std::string rotorTypeList[] = {"Asymmetric",
                                     "Symmetric",
                                     "Spherical",
                                     "Linear",
                                     "Atom"};

/*! \file Molecule.hpp
 *  \class Molecule
 *  \brief Class representing a molecule or general aggregate of atoms.
 *  \author Roberto Di Remigio
 *  \date 2014
 *
 *  This class is based on the similar class available in the Mints library
 *  of Psi4
 */

class Molecule {
/* Class members and methods here */
};
/*! @} End of Doxygen Groups*/
```

**Comment by arnfinn** [comment]
Can you give me an example on how to do it the `Atom.hpp` file? If I do `.. doxygenstruct:: pcm::utils::Atom` and move
```
/*! \file Atom.hpp
 *  \struct Atom
 *  \brief A POD describing an atom.
 *  \author Roberto Di Remigio
 *  \date 2011, 2016
 */
```
inside namespace and
```
/*!
 *  \addtogroup pcm::utils
 *  @{
  */
``` 
outside namespace, I do get this:
![screen shot 2017-10-25 at 15 59 18](https://user-images.githubusercontent.com/136346/32002748-81e709e8-b99d-11e7-9c4c-a18094fd3bc4.png)

Thus, without the \brief text, author etc.

**Comment by robertodr** [comment]
I think we need to rewrite **all** of the Doxygen docstrings :sweat_smile:
We want to have:
1. `\file` at the top.
2. `\addtogroup` before any `namespace`.
3. `\class` or `\struct` before any class or struct.

For `Atom.hpp` it would look like this:
```
#pragma once

#include <iosfwd>
#include <string>
#include <vector>

#include "Config.hpp"

#include <Eigen/Core>

/*! \file Atom.hpp */

/*!
 *  \addtogroup pcm
 *  @{
 */
namespace pcm {
template <typename CreateObject> class Factory;

/*!
 *  \addtogroup utils
 *  @{
 */
namespace utils {
/*! \struct Atom
 *  \brief A POD describing an atom.
 *  \author Roberto Di Remigio
 *  \date 2011, 2016
 */
struct Atom {
  Atom()
      : charge(0.0),
        mass(0.0),
        radius(0.0),
        radiusScaling(0.0),
        position(Eigen::Vector3d::Zero()),
        element("Dummy"),
        symbol("Du") {}
  Atom(const std::string & elem,
       const std::string & sym,
       double c,
       double m,
       double r,
       const Eigen::Vector3d & coord,
       double scal = 1.0)
      : charge(c),
        mass(m),
        radius(r),
        radiusScaling(scal),
        position(coord),
        element(elem),
        symbol(sym) {}
  friend std::ostream & operator<<(std::ostream & os, Atom & at) {
    os << "Atom: " << at.symbol << " " << at.charge << " " << at.mass << std::endl;
    os << "  Radius " << at.radius << std::endl;
    os << "  Position " << at.position.transpose();
    return os;
  }
  /*! Atomic charge */
  double charge;
  /*! Atomic mass */
  double mass;
  /*! Atomic radius */
  double radius;
  /*! Scaling of the atomic radius */
  double radiusScaling;
  /*! Position of the atom */
  Eigen::Vector3d position;
  /*! Name of the element */
  std::string element;
  /*! Atomic symbol */
  std::string symbol;
  EIGEN_MAKE_ALIGNED_OPERATOR_NEW /* See
                                     http://eigen.tuxfamily.org/dox/group__TopicStructHavingEigenMembers.html
                                     */
};

typedef pcm::tuple<std::string, std::vector<Atom> > RadiiSet;

/*!
 *  \addtogroup utils
 *  @{
 */
namespace detail {
/*! \brief Returns a vector<Atom> containing Bondi van der Waals
 *radii.
 *
 * The van der Waals radii are taken from:
 * --- A. Bondi, J. Phys. Chem. 68, 441-451 (1964)
 * complemented with the ones reported in:
 * --- M. Mantina, A. C. Chamberlin, R. Valero, C. J. Cramer, D. G. Truhlar,
 *     J. Phys. Chem. A, 113, 5806-5812 (2009)
 * We are here using Angstrom as in the papers.
 */
RadiiSet initBondi();

/*! \brief Returns a vector<Atom> containing UFF radii.
 *
 * The UFF set of radii is taken from:
 * --- A. Rappé, C. J. Casewit, K. S. Colwell, W. A. Goddard, W. M. Skiff,
 *     J. Am. Chem. Soc., 114, 10024-10035 (1992)
 * We are here using Angstrom as in the paper.
 */
RadiiSet initUFF();

/*! \brief Returns a vector<Atom> containing Allinger's MM3 radii.
 *
 * The MM3 set of radii is taken from:
 * --- N. L. Allinger, X. Zhou, J. Bergsma,
 *     J. Mol. Struct. (THEOCHEM), 312, 69-83 (1994)
 * We are here using Angstrom as in the paper.
 *
 * \note We *divide* the values reported in the paper by 1.2, as done in
 * the ADF program package.
 */
RadiiSet initAllinger();

typedef pcm::function<RadiiSet()> CreateRadiiSet;
} // namespace detail
/*! @} */

Factory<detail::CreateRadiiSet> bootstrapRadiiSet();
} // namespace utils
/*! @} */

/*! An atom is invalid if it has zero radius */
bool invalid(const utils::Atom & atom);
} // namespace pcm
/*! @} */
```

**Comment by arnfinn** [comment]
Closed by #102
